Add support for KV.txn and add a Lock class

ducdigital opened the pull request:
- [x] Add support for all compare ops
  - [x] CompareValue
  - [x] CompareLastRevision
  - [x] CompareCreated
  - [x] CompareVersion
- [x] Add support for all success/failure request
  - [x] DeleteRangeRequest
  - [x] RangeRequest
  - [x] PutRequest


Example usage:
```
  let compareOps = [
    new CompareValue(
      CompareTypeDef.CompareResult.EQUAL,
      new Buffer('key-to-compare'),
      new Buffer('key-to-compare')
    )
  ];
  let successOps = [
    new PutRequest({
      key: new Buffer('put-if-true-key'),
      value: new Buffer('put-if-true-value'),
      lease: 0,
    })
  ];
  let failureOps = [
    new PutRequest({
      key: new Buffer('put-if-fail-key1'),
      value: new Buffer('put-if-fail-value1'),
      lease: 0,
    }),
    new PutRequest({
      key: new Buffer('put-if-fail-key2'),
      value: new Buffer('put-if-fail-value2'),
      lease: 0,
    })
  ];

  etcd.createTransaction(compareOps, successOps, failureOps).then(data => {
    console.log(data);
  });
```

PS: This is experimental just to kick start the idea. It works but the code is not as clean as the original code base. Might need some refactoring work.

ducdigital committed:
Add support for KV.txn

CitySim commented:
hi,
thanks for the pull request! 👍 
I'll make sure to check it out this week

ducdigital committed:
Update Proto file to latest

ducdigital committed:
Update core transaction and up to es6

The problem with etcd proto is that if we use it's class in an OOP way
it does not really work.

ducdigital committed:
ETCD Lock implementation

ducdigital committed:
Merge branch 'feature/kv-txn-lock' into feature/kv-txn
Add Lock functionality
Fix problem of transaction cause of using nested classes of proto object. it does not really work

ducdigital commented:
Hey there I just update the Transaction class and add a Lock class. Lock class is "almost" direct implementation of the Python's version.

I also update Transaction class to fix the problem cause by nested proto class object (It does not really work that way and I am not sure why). 

As soon as I change everything to a normal object It work very well.

ducdigital committed:
Minor update add Watch service
